appmap/flask.py

```diff
@@ -100,46 +106,87 @@ def record_delete(self):
         return json.loads(generation.dump(self.recording))
 
     def before_request(self):
-        if self.recording.is_running() and request.path != self.record_url:
-            Metadata.add_framework("flask", flask.__version__)
-            np = None
-            # See
-            # https://github.com/pallets/werkzeug/blob/2.0.0/src/werkzeug/routing.py#L213
-            # for a description of parse_rule.
-            if request.url_rule:
-                np = "".join(
-                    [
-                        f"{{{p}}}" if c else p
-                        for c, _, p in parse_rule(request.url_rule.rule)
-                    ]
-                )
-            call_event = HttpServerRequestEvent(
-                request_method=request.method,
-                path_info=request.path,
-                message_parameters=request_params(request),
-                normalized_path_info=np,
-                protocol=request.environ.get("SERVER_PROTOCOL"),
-                headers=request.headers,
-            )
-            Recorder.add_event(call_event)
-
-            appctx = _app_ctx_stack.top
-            appctx.appmap_request_event = call_event
-            appctx.appmap_request_start = time.monotonic()
+        if Env.current.enabled or self.recording.is_running():
+            # there are three options: remote record_requests both
+            if not Env.current.record_all_requests and self.recording.is_running():
+                self.before_request_main([Recorder()])
+            else:
+                rec = Recorder(_EventIds.get_thread_id())
+                rec.start_recording()
+                recorders = [rec]
+                if self.recording.is_running():
+                    recorders.append(Recorder())
```

Comment:
It's different because Flask makes available [two handlers](https://github.com/getappmap/appmap-python/blob/sw/feat/record_requests__version3/appmap/flask.py#L87) for the request, one before and one after, instead of a single handler.

I like the idea of factoring it out into web_framework so the two frameworks can share it.  I'll try this.

Reply:
Added tests for "remote recording while per-request recording".

Did you mean to also add tests for "remote recording while not per-request recording"?

Reply:
No, we already have those.